src/Facets/CalldataVerificationFacet.sol

```diff
@@ -0,0 +1,240 @@
+    function extractBridgeData(
+        bytes calldata data
+    ) external pure returns (ILiFi.BridgeData memory bridgeData) {
+        bridgeData = abi.decode(data[4:], (ILiFi.BridgeData));
+        return bridgeData;
+    }
+
+    /// @notice Extracts the swap data from the calldata
+    /// @param data The calldata to extract the swap data from
+    /// @return swapData The swap data extracted from the calldata
+    function extractSwapData(
+        bytes calldata data
+    ) external pure returns (LibSwap.SwapData[] memory swapData) {
+        (, swapData) = abi.decode(
+            data[4:],
+            (ILiFi.BridgeData, LibSwap.SwapData[])
+        );
+        return swapData;
+    }
+
+    /// @notice Extracts the bridge data and swap data from the calldata
+    /// @param data The calldata to extract the bridge data and swap data from
+    /// @return bridgeData The bridge data extracted from the calldata
+    /// @return swapData The swap data extracted from the calldata
+    function extractData(
+        bytes calldata data
+    )
+        external
+        pure
+        returns (
+            ILiFi.BridgeData memory bridgeData,
+            LibSwap.SwapData[] memory swapData
+        )
+    {
+        (bridgeData) = abi.decode(data[4:], (ILiFi.BridgeData));
+        if (bridgeData.hasSourceSwaps) {
+            (, swapData) = abi.decode(
+                data[4:],
+                (ILiFi.BridgeData, LibSwap.SwapData[])
+            );
+        }
+
+        return (bridgeData, swapData);
+    }
+
+    /// @notice Extracts the main parameters from the calldata
+    /// @param data The calldata to extract the main parameters from
+    /// @return bridge The bridge extracted from the calldata
+    /// @return sendingAssetId The sending asset id extracted from the calldata
+    /// @return receiver The receiver extracted from the calldata
+    /// @return minAmount The min amount extracted from the calldata
+    /// @return destinationChainId The destination chain id extracted from the calldata
+    /// @return hasSourceSwaps Whether the calldata has source swaps
+    /// @return hasDestinationCall Whether the calldata has a destination call
+    function extractMainParameters(
+        bytes calldata data
+    )
+        public
+        pure
+        returns (
+            string memory bridge,
+            address sendingAssetId,
+            address receiver,
+            uint256 minAmount,
+            uint256 destinationChainId,
+            bool hasSourceSwaps,
+            bool hasDestinationCall
+        )
+    {
+        ILiFi.BridgeData memory bridgeData;
+        (bridgeData) = abi.decode(data[4:], (ILiFi.BridgeData));
+
+        if (bridgeData.hasSourceSwaps) {
+            LibSwap.SwapData[] memory swapData;
+            (, swapData) = abi.decode(
+                data[4:],
+                (ILiFi.BridgeData, LibSwap.SwapData[])
+            );
+            sendingAssetId = swapData[0].sendingAssetId;
```

Comment:
I think we always return the `bridgeData.sendingAssetId` for the `sendingAssetId`.
When we do swaps before the bridging, the final asset of swaps is `sendingAssetId`.

Reply:
Yes, but we want to return which asset the user starts with in this case.